Solutions/1.2-Introduction-to-Musig-Solutions.ipynb

```diff
@@ -0,0 +1,212 @@
+    "pk1_c = pk1.mul(c_map[pk1])\n",
+    "pk2_c = pk2.mul(c_map[pk2])\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 2: Nonce Generation\n",
+    "\n",
+    "* Nonces can be aggregated like public keys and signatures.\n",
+    "* Security [Proof](https://eprint.iacr.org/2018/068.pdf) requires nonces are randomly generated independently of each other.\n",
+    "* An initial nonce commitment round is necessary to ensure no nonce is derived as a function of another.\n",
+    "\n",
+    "\n",
+    "![test](../images/musig_intro_2.jpg)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.1 Programming Exercise: Compute 3-of-3 musig nonce.\n",
+    "\n",
+    "* `generate_schnorr_nonce()` returns a ECkey with with a jacobi symbol of 1.\n",
+    "* `aggregate_schnorr_nonces()` returns a ECPubKey and a boolean (indicates nonce negation).\n",
+    "* *Negation of nonce `k_i` isn't necessary, since it is performed by `sign_musig` in next exercise.*"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate nonces.\n",
+    "k0 = generate_schnorr_nonce()\n",
+    "k1 = generate_schnorr_nonce()\n",
+    "k2 = generate_schnorr_nonce()\n",
+    "R0 = k0.get_pubkey()\n",
+    "R1 = k1.get_pubkey()\n",
+    "R2 = k2.get_pubkey()\n",
+    "\n",
+    "# Skip nonce point commitment round (we assume nonce commitments are correct).\n",
+    "\n",
+    "# Aggregate nonces.\n",
+    "R_agg, negated = aggregate_schnorr_nonces([R0, R1, R2])\n",
+    "print(\"R aggregated: \", R_agg.get_bytes().hex())\n",
+    "print(\"R_agg was negated: \", negated)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 3: Signature Generation\n",
+    "\n",
+    "* Signatures are generated individually.\n",
+    "\n",
+    "![test](images/musig_intro_3.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.1 Programming exercise: Compute aggregated musig signature.\n",
+    "\n",
+    "* `ECKey.sign_musig(nonce_point, negated, nonce_point_aggregated, pk_musig, msg)` returns individual signature.\n",
+    "* `aggregate_musig_signatures([sig0, sig1, ...])` returns aggregated signature.\n",
+    "* `ECPubKey.verify(sig, msg)` returns validity of signature.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "msg = hashlib.sha256(b'transaction').digest()\n",
+    "\n",
+    "# Generate individual signatures.\n",
+    "sig0 = sec0_c.sign_musig(k0, negated, R_agg, pk_musig, msg)\n",
+    "sig1 = sec1_c.sign_musig(k1, negated, R_agg, pk_musig, msg)\n",
+    "sig2 = sec2_c.sign_musig(k2, negated, R_agg, pk_musig, msg)\n",
+    "\n",
+    "# Aggregate signatures.\n",
+    "sig_agg = aggregate_musig_signatures([sig0, sig1, sig2])\n",
```

Comment:
It'd be nice if you could make it much more explicit that the signature generated using the MuSig validates correctly.